### ISSUE TITLE
Set data at *physical* row instead of visual.

### ISSUE THREAD

**Issue by aarondfrancis** [handsontable/handsontable#6309]
Hey there,

I'm running into a real issue trying to set some data because the `setDataAtRowProp` and `setDataAtCell` both only accept visual row indexes and not physical:

https://github.com/handsontable/handsontable/blob/98878a6bb6edbed065b7961bc97d4b799e908503/src/core.js#L1215

As far as I can tell, there is no way to set data using a physical row index.

What I'm running into is a scenario when some rows have been trimmed (so they have no visual index) and then data comes in from the server that needs to update one of those trimmed rows.

Can we get new methods to set values by physical indexes? And in the meantime, what is the suggested workaround? 

Thanks!

**Comment by AMBudnik** [comment]
Thank you for sharing this issue Aaron.

I have built a small example to replicate the issue. https://jsfiddle.net/fcd493sj/ The version loaded is `8.0.0rc1 build 26/09/19` but in this case, it doesn't differ from `7.1.1`

### Scenario 1

![Zrzut ekranu 2019-10-4 o 10 00 03](https://user-images.githubusercontent.com/10757813/66191211-185cf300-e68e-11e9-959a-cce875301b51.png)

I'm trimming row index `0` and use 

```js
hot.setDataAtRowProp(0, 'name', 'Maria');
```

to change `Alex` to `Maria`.

**Result**
![Zrzut ekranu 2019-10-4 o 09 53 53](https://user-images.githubusercontent.com/10757813/66191271-42161a00-e68e-11e9-8cba-43e1eebf73c3.png)

### Scenario 2

Now let's try with 

```js
`hot.setDataAtRowProp(hot.toPhysicalRow(0),` 'name', 'Maria');
```

**Result**
![Zrzut ekranu 2019-10-4 o 09 53 35](https://user-images.githubusercontent.com/10757813/66191387-7ee21100-e68e-11e9-81ef-d52c4b8475c7.png)

### Conclusion
After seeing `Maria` at the last position I came to the conclusion that this looks exactly like the `alter` issue I've posted recently. 
Here https://github.com/handsontable/handsontable/issues/6307 we have the same situation when the visual index is gone so in the result data is added to a wrong index.

Alternatively, we could add a `setSourceDataAtCell` method that would replace values based on `getSourceData`. As in the example above `getSourceData` returns the whole dataset.

![Zrzut ekranu 2019-10-4 o 10 12 16](https://user-images.githubusercontent.com/10757813/66191823-776f3780-e68f-11e9-8f50-c0a0363d5b39.png)

**Comment by aarondfrancis** [comment]
@AMBudnik thanks for the thorough explanation!

As for: 
> Alternatively, we could add a setSourceDataAtCell method

that would be 100% my preference. I often am trying to just set the underlying source data and end up having to do a lot of runaround with translating row indexes.

**Comment by aarondfrancis** [comment]
For anyone following, here's what I ended up with.

```js
withoutTrimmedRows(cb) {
    const alreadyTrimmed = plugins.trim.getTrimConfig();

    // If no rows are trimmed, we don't need to waste any
    // time untrimming and re-trimming.
    if (alreadyTrimmed.length) {
        plugins.trim.untrimAll();
    }

    cb();

    if (alreadyTrimmed.length) {
        plugins.trim.trimRows(alreadyTrimmed);
        
        // The user supplied callback could have called render(),
        // which would mean that the rows we untrimmed are now
        // visible. So here we'll re-render so that the trimmed
        // rows disappear again.
        hot.render();
    }
},

applyChangesFromServer(changes) {
    this.withoutTrimmedRows(() => {
        // Apply whatever changes you want, knowing that 
        // all physical rows will have visual indexes.
    });
} 
```

It's pretty straightforward, but what it does is untrim everything, call your callback where you make your changes, and then retrim the rows. 

Works well enough for now, until we can set data at physical indexes!

**Comment by aarondfrancis** [comment]
For anyone following along... 

The code sample I pasted above worked really well for trimmed rows, but it fell flat when it came to rows that the user had filtered using the Filters plugin.

I've got a basic little plugin now that adds a new `setDataAtPhysicalRowProp` method onto my HoT instance. I'll try to package it up and put it on npm at some point.

Be warned, this is kind of crazy and I've only tested it on 8.0.0. I know a lot of work has gone into the record translator layer for 8.0, so this probably won't work on 7.x.

Because the core `setDataAtRowProp` [calls the translator's `toPhysicalRow` method](https://github.com/handsontable/handsontable/blob/98878a6bb6edbed065b7961bc97d4b799e908503/src/core.js#L1215), this plugin hijacks that method and lets you return whatever you want.

So knowing that, the `setDataAtPhysicalRowProp` will mutate the row number into a recognizable pattern, and our hijacked `toPhysicalRow` will reverse it.

Like I said, kinda crazy, but it makes my life way way easier. Now I can set data by physical index! 

Looking forward to a non-crazy way to do this in core, but for now this is totally passable.

```js
import {getTranslator} from 'handsontable/commonjs/translations/recordTranslator';

// Omitted for brevity: setting up PhysicalIndexes as a proper HoT plugin

PhysicalIndexes.prototype.enable = function () {
    const translator = getTranslator(this.hot);
    const originalToPhysicalRow = translator.toPhysicalRow;
    const plugin = this;

    translator.toPhysicalRow = function (row) {
        // If it's mutated, then it's one of ours and we should
        // unmutate it and return, bypassing the core method.
        if (plugin.isMutated(row)) {
            return plugin.unmutate(row);
        }

        // If it's not mutated, then just pass it along.
        return originalToPhysicalRow.call(this, row);
    }

    this.hot.setDataAtPhysicalRowProp = function (row, prop, value, source) {
        // Mutate the row numbers. The first argument could be an
        // array, indicating that it is a set of changes rather
        // than just a single change.
        if (typeof row === 'object') {
            for (let i = 0; i < row.length; i++) {
                row[i][0] = plugin.mutate(row[i][0]);
            }
        } else {
            row = plugin.mutate(row);
        }

        this.setDataAtRowProp(row, prop, value, source);
    }
}

PhysicalIndexes.prototype.isMutated = function (row) {
    return row < 0;
}

PhysicalIndexes.prototype.mutate = function (row) {
    // We can't just turn it negative, because that leaves 0 as ambiguous.
    // So we'll first add 1, and then turn it negative.
    return -(row + 1);
}

PhysicalIndexes.prototype.unmutate = function (row) {
    // First flip the sign back, then remove our extra 1.
    return -row - 1;
}

export default PhysicalIndexes;
```

**Comment by AMBudnik** [comment]
I enjoy watching how 'mad scientists' are hacking Handsontable. It sounds like a complex but reliable plan. I especially like the idea that you've made it as a plugin.

I hope that we'd be able to push `setDataAtPhysicalRowProp`/`setDataAtPhysicalRow` once we get rid of all the indexing issues. Index mapper is being tested and it goes well. We should be able to publish it at the beginning of November.

**Comment by OlgaIOn** [comment]
We are also running into the same issue. Is there any update on this? When will be setDataAtPhysicalRowProp / setDataAtPhysicalRow available?

**Comment by AMBudnik** [comment]
We're pretty close to reaching this goal. Here's a related topic https://github.com/handsontable/handsontable/issues/6546

**Comment by wojciechczerniak** [comment]
`v8.0.0` from `beta.2` introduces `setSourceDataAtCell`. It includes "atRowProp", we've decided to not introduce two new methods since at the Source Data level they're practically the same.
https://github.com/handsontable/handsontable/blob/6d24b71d46041ad0bcbfcbddc7b8ba6550823382/src/core.js#L2423-L2430

New hooks were introduced as well so we can attach a middleware callback and modify the source:
https://github.com/handsontable/handsontable/blob/6d24b71d46041ad0bcbfcbddc7b8ba6550823382/src/pluginHooks.js#L1005-L1015

Or respond to changes:
https://github.com/handsontable/handsontable/blob/6d24b71d46041ad0bcbfcbddc7b8ba6550823382/src/pluginHooks.js#L588-L596

As mentioned in https://github.com/handsontable/handsontable/issues/6309#issuecomment-538441112 those are the desired solution to this issue.

**Comment by aarondfrancis** [comment]
Excellent, thanks so much @wojciechczerniak and team! I'll update and report back. I'll be happy to remove my wild workarounds.

**Comment by wojciechczerniak** [comment]
Great! It should be available from NPM on Wednesday

**Comment by pnowak** [comment]
Hi @aarondfrancis @OlgaIOn 

I have good news! We've just released 8.0.0-beta2 that fixes the mentioned issue. We won't be closing this ticket now. It will be closed after we release the final 8.0.0 version of Handsontable.

Here’s a full list of changes https://github.com/handsontable/handsontable/releases/tag/8.0.0-beta.2
NPM https://www.npmjs.com/package/handsontable/v/8.0.0-beta.2
 
All feedback's appreciated. Thank you for your input.

**Comment by AMBudnik** [comment]
Hi @aarondfrancis 

I’m glad to announce that Handsontable v8.0.0 that solves this issue is live!   

CDN https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.css
https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.js
NPM npm i handsontable
Release Notes https://github.com/handsontable/handsontable/releases  

I highly encourage you to test the new version and leave some feedback. ✍️  I hope that it will improve your project :)
